x/milestone/abci/abci.go

```diff
@@ -0,0 +1,315 @@
+	var supportingValidatorList []string
+	for valAddr, blocks := range validatorVotes {
+		supports := true
+		for blockNum := startBlock; blockNum <= endBlock; blockNum++ {
+			hash, hasBlock := blocks[blockNum]
+			if !hasBlock || !bytes.Equal(hash, blockToHash[blockNum]) {
+				supports = false
+				break
+			}
+		}
+		if supports {
+			supportingValidatorList = append(supportingValidatorList, valAddr)
+		}
+	}
+
+	// Sort validators deterministically
+	sort.Strings(supportingValidatorList)
+
+	// Verify that we still have 2/3 majority after filtering
+	totalSupportingPower := int64(0)
+	for _, valAddr := range supportingValidatorList {
+		totalSupportingPower += valAddressToVotingPower[valAddr]
+	}
+
+	if totalSupportingPower < majorityVP {
+		logger.Debug("After filtering validators, no range has 2/3 majority support",
+			"totalSupportingPower", totalSupportingPower,
+			"requiredPower", majorityVP)
+		return nil, nil, "", nil
+	}
+
+	// Additional sort by voting power (stable to preserve string order when tied)
+	sort.SliceStable(supportingValidatorList, func(i, j int) bool {
+		return valAddressToVotingPower[supportingValidatorList[i]] > valAddressToVotingPower[supportingValidatorList[j]]
+	})
+
+	if len(supportingValidatorList) == 0 {
+		return nil, nil, "", fmt.Errorf("no validators support the winning range")
+	}
+
+	// Generate aggregated proposers hash from supporting validators
+	aggregatedProposersHash := []byte{}
+	for _, valAddr := range supportingValidatorList {
+		aggregatedProposersHash = crypto.Keccak256(
+			aggregatedProposersHash,
+			[]byte{'|'},
+			validatorAddresses[valAddr],
+		)
+	}
+
+	// Create final proposition
+	proposition := &types.MilestoneProposition{
+		BlockHashes:      blockHashes,
+		StartBlockNumber: startBlock,
+	}
+
+	logger.Debug("Found majority milestone proposition",
+		"startBlock", startBlock,
+		"endBlock", endBlock,
+		"blockCount", blockCount,
+		"supportingValidators", len(supportingValidatorList))
+
+	return proposition, aggregatedProposersHash, supportingValidatorList[0], nil
+}
+
+func getBlockHashes(ctx sdk.Context, startBlock, maxBlocksInProposition uint64, contractCaller helper.IContractCaller) ([][]byte, error) {
+	result := make([][]byte, 0)
+
+	headers, err := contractCaller.GetBorChainBlocksInBatch(ctx, int64(startBlock), int64(startBlock+maxBlocksInProposition-1))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get headers")
+	}
+
+	for _, h := range headers {
+		result = append(result, h.Hash().Bytes())
+	}
+
+	return result, nil
+}
+
+func ValidateMilestoneProposition(ctx sdk.Context, milestoneKeeper *keeper.Keeper, milestoneProp *types.MilestoneProposition) error {
+	if milestoneProp == nil {
+		return nil
+	}
+
+	params, err := milestoneKeeper.GetParams(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(milestoneProp.BlockHashes) > int(params.MaxMilestonePropositionLength) {
+		return fmt.Errorf("too many blocks in proposition")
+	}
```

Comment:
Another thought... If there is a limit on the size of vote extension message, we would already have an implicit limit on the number of blocks a validator can put into a milestone prop. In that case, do we still need this validation?

Reply:
If the concern is about the calculation time, we can still use `MaxMilestonePropositionLength` and put a hard stop in `GetMajorityMilestoneProposition`.